### DISCUSSION TITLE
Ruby on rails

### DISCUSSION THREAD

**Discussion by B3Y0N4** [19/classroom#7] [Ruby]
Ruby on rails bende yüklü olduğu halde yeni bir dosya kurmaya çalışırken böyle bir hata alıyorum bash: new: command not found 
![rails hata](https://user-images.githubusercontent.com/72824981/152781037-8cf3a1af-2971-4c68-9618-e4c542ccc3f4.jpg)

**Comment by roktas** [comment]
Doğal Windows ortamında Rails geliştirmesi yapmak vakit kaybından başka bir şey değil maalesef.  Bir süre bekle, Windows üzerinde WSL yoluyla kurulan bir Linux ortamında geliştirme yapmanız için bir hazırlığım var.  Soruyu şimdilik "cevaplanmamış" bırakıyorum.  Bahsettiğim çalışma tamamlanınca duyuracağım.

**Reply by BunyaminAkdeniz** [reply]
Hocam Ruby and Rails'a başlamayı istiyorum. Nereden başlamalıyım, nereden öğrenebilirim. Önerebileceğiniz bir kaynak var mıdır ?  @roktas

**Reply by roktas** [reply]
Bir numaralı kaynak Rails'ın kendi dokümanları: https://guides.rubyonrails.org/

**Reply by BunyaminAkdeniz** [reply]
Merhaba Hocam @roktas . Rails ile çalışırken şöyle bir durum fark ettim. Oluşturduğum Rails uygulamasında şöyle bir dosya bulunuyor. `app/controllers/articles_controller.rb`. Ve içeriği şu şekilde 

> class ArticlesController < ApplicationController
>    def index
>    end
> end

Fakat bu dosyanın üstünde bir `require` ya da `require_relative` yok. Bu kod nasıl hatasız çalışıyor ?

**Reply by roktas** [reply]
Kısa cevap "black magic" veya "rails magic" :-)  Uzun cevap: Rails'da gerekli noktalarda gerekli "sabitlerin" (sınıf, modül, sabit) yüklenmesi için bir mekanizma var.  Bu mekanizmanın bir tarafında Rails konvansiyonları (bk. ["Convention over Configuration"](https://rubyonrails.org/doctrine#convention-over-configuration)) diğer tarafında özel bir kitaplık (bk. [Zeitwerk](https://github.com/fxn/zeitwerk)) var.  Parantez içindeki anahtar terimleri aratmanı öneririm.  Öte yandan Rails dokümantasyonundan nokta atışı yapacak olursak: https://guides.rubyonrails.org/autoloading_and_reloading_constants.html
  
Özetle iş "Zeitwerk"de bitiyor ve bunun da yaptığı kabaca `require`'ı (ki Ruby'de bu bir Kernel metodudur; Ruby'de dile ait hemen her şeyi overwrite edebilirsin) tekrar tanımlamak.  Bu sayede kendi tanımladığın `require` ile kaynak koddaki tüm `require` çağrılarını "intercept" edebiliyorsun (yani araya girerek "üç kaat" yapabiliyorsun).

**Reply by BunyaminAkdeniz** [reply]
Hocam Ruby başlarda çok doğal ve anlaşılır bir dildi ancak derine indikçe çok daha karmaşık bir hal alıyor gibi. Yanıtınız için teşekkür ederim. Son editiniz ile çok daha iyi anladım. Bu sizin derste söylediğiniz monkey patching'e örnek midir ?

**Reply by roktas** [reply]
Evet, modül düzeyinde yapılan bir "monkey patching".  Ruby konusundaki yargın doğru.  Ruby basit gibi görünen fakat ustalaşması zor olan bir dildir.  Yolun başında dikkat etmeni önereceğim bir konu da "Rails etkisi".  Rails bir bakıma ustalaşması zor olan Ruby'de çok fazla uğraşmadan çabuk sonuçlar almanı sağlayan bir framework.  Ama bu seni yanıltmasın.  Ruby dünyası Rails'dan ibaret değil ve her zaman öğütlediğim konu, yol üzerinde Ruby ustalığınızı arttırın.  Buna benzer görüşlere [kulak verelim](https://solnic.codes/2022/02/02/rails-and-its-ruby-dialect/).
